src/Admin/AbstractAdmin.php

```diff
@@ -15,6 +15,30 @@
 use Knp\Menu\FactoryInterface as MenuFactoryInterface;
 use Knp\Menu\ItemInterface;
 use Knp\Menu\ItemInterface as MenuItemInterface;
+use Sonata\AdminBundle\Admin\AdminExtensionInterface as OldAdminExtensionInterface;
+use Sonata\AdminBundle\Admin\Extension\AdminExtensionInterface as NewAdminExtensionInterface;
+use Sonata\AdminBundle\Admin\Extension\AlterNewInstanceInterface;
+use Sonata\AdminBundle\Admin\Extension\AlterObjectInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureActionButtonsInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureBatchActionsInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureDatagridFieldsInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureDefaultFilterValuesInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureExportFieldsInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureFormFieldsInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureListFieldsInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureQueryInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureRoutesInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureShowFieldsInterface;
+use Sonata\AdminBundle\Admin\Extension\ConfigureTabMenuInterface;
+use Sonata\AdminBundle\Admin\Extension\GetAccessMappingInterface;
+use Sonata\AdminBundle\Admin\Extension\GetPersistentParametersInterface;
+use Sonata\AdminBundle\Admin\Extension\PostPersistInterface;
+use Sonata\AdminBundle\Admin\Extension\PostRemoveInterface;
+use Sonata\AdminBundle\Admin\Extension\PostUpdateInterface;
+use Sonata\AdminBundle\Admin\Extension\PrePersistInterface;
+use Sonata\AdminBundle\Admin\Extension\PreRemoveInterface;
+use Sonata\AdminBundle\Admin\Extension\PreUpdateInterface;
+use Sonata\AdminBundle\Admin\Extension\ValidateInterface;
```

Comment:
Holy shit… maybe introduce a new component responsible for calling the extensions? And then call it like this : `$this->component->fireEvent(Component::ALTER_NEW_INSTANCE, $this);`?

Reply:
If the use statements are problem I can do `use Sonata\AdminBundle\Admin\Extension;` and then call `Extension\PreUpdateInterface`.

About your suggestion, replacing:

```
foreach ($this->getExtensions() as $extension) {
            if ($extension instanceof ConfigureExportFieldsInterface) {
                $fields = $extension->configureExportFields($this, $fields);
            }
        }
```

with `$this->component->fireEvent(Component::CONFIGURE_EXPORT_FIELDS, $this);`?

Component would be some kind of Registry with all possible interfaces for extensions and then it would be something like:

```
foreach ($this->getExtensions() as $extension) {
            if ($extension instanceof $this->interfaces[$event]) {
                 $extension->configureExportFields($this, $fields);
            }
        }
```

The problem I see is that some of those fields/methods are private/protected or even in the scope of that method. Any solution for that? Don't think passing `$this` with the event would help, maybe introducing an DTO that would be used instead of $this?

Reply:
Basically we would be building something like `symfony/event-dispatcher`?

Reply:
> The problem I see is that some of those fields/methods are private/protected or even in the scope of that method.

Can you give an example of this, I don't follow?

Reply:
> Basically, we would be building something like symfony/event-dispatcher?

Not quite, since there wouldn't be subscribers (for several events), only listeners (for only one event)

Reply:
I explained badly, I was thinking about 

```
$extension->postPersist($this, $object);
$extension->configureTabMenu($this, $menu, $action, $childAdmin);
$extension->configureActionButtons($this, $list, $action, $object);
```

Reply:
Side note, there is a PSR in the works for this https://github.com/php-fig/fig-standards/blob/master/proposed/event-dispatcher.md

And maybe being able to also have subscribers, not just listeners could be good. And what we have currently looks a lot like an event dispatching system.

Reply:
We could just have 

`$this->component->fireEvent(Component::CONFIGURE_EXPORT_FIELDS, [$this, $menu, $action, $childAdmin]);`

And just forward that arguments as `...$arguments`?

Reply:
Not sure if I should close this PR then and start a new one?

Reply:
What about I fix other stuff you suggested and start another PR but branch it from this branch and then when the 2nd one is done, we merge this one and I rebase and we merge the second one?

Reply:
Good idea! Mark this one as WIP please. I'm going to read the PSR, there's a lot of interesting vocabulary in here, for instance, it looks like our component is a "notifier", because it knows the listener whereas the admin does not. The admin is an emitter.

Reply:
Or maybe the component is a listener provider, I'm not sure yet, let's read :eyeglasses:

Reply:
I just read the PSR and a bit confused, since some extensions methods return values we would need Messages + Task, Emitter and Listener Provider. Extensions would be our Listeners. Not sure if we need Processor + Notifier, it looks a bit too complicated for our case, can't we just emit the event to listener provider and he forwards it where it needs to go?

Reply:
IIRC you can have the Notifier and Listener be the same class, so yes, we can.

Reply:
I misunderstood, I think I got it now:

We have Events (Message/Task) that are emitted with Emitter to Notifier/Processor, that then get a list of Listeners from Listener Provider based on the Event, the only difference is that Processor returns value while Notifier doesn't.